### PR TITLE
Fix zarr 3.0 compatibility

### LINKED ISSUE #7193: [test-bot] pip install --pre is failing
The --pre Test workflow failed on 2024-08-31 12:21 UTC

The most recent failing test was on windows-latest py3.12 pyqt6
with commit: 6e292a8a05f2d195882288d4629695047896f20c

Full run: https://github.com/napari/napari/actions/runs/10645077004

(This post will be updated if another test fails, as long as this issue remains open.)

### LINKED ISSUE #7279: [test-bot] pip install --pre is failing
The --pre Test workflow failed on 2024-11-06 12:22 UTC

The most recent failing test was on windows-latest py3.12 pyqt5
with commit: 2f6addc9454127143fe1236248ca82bc6d95dfad

Full run: https://github.com/napari/napari/actions/runs/11703223331

(This post will be updated if another test fails, as long as this issue remains open.)

### PR DESCRIPTION
**Pull request by Czaki**
# References and relevant issues
closes #7279

# Description

Fix zarr 3.0 related test by used named arguments

### CODE REVIEW

**Comment by psobolewskiPhD** [comment]
One of the big  changes of Zarr v3 is that everything now goes in zarr.json instead of .zarray + .zgroup + .zattrs

**Review by psobolewskiPhD — DISMISSED** [review]
One of the big changes of Zarr v3 is that everything now goes in zarr.json instead of .zarray + .zgroup + .zattrs so need to update the logic in builtins to use either or.

**Comment by Czaki** [comment]
It looks also like `zarr.open` is a problem as it does not allow creating group file.

**Comment by psobolewskiPhD** [comment]
Looks like need to block `pydantic==2.9.0b1` for this PR.

**Comment by Czaki** [comment]
There is even issue for this https://github.com/pydantic/pydantic/issues/10239 initially reported here https://x.com/Czaki_PL/status/1828178992332955873

File: napari_builtins/io/_read.py [hunk 1]
```diff
@@ -133,6 +133,12 @@ def read_zarr_dataset(path: str):
         ]
         assert image, 'No arrays found in zarr group'
         shape = image[0].shape
+    elif (path / 'zarr.json').exists():
+        # zarr v3
+        import zarr
+
+        image = zarr.open(store=path)
+        shape = image.shape
```
Comment (jni) [inline napari_builtins/io/_read.py]: @Czaki this use case probably also needs to be split into single array/group of arrays? I presume that we can make the earlier ifs more complex: "if there's a .zarray *or* there's a zarr.json and it only contains one array"

Comment (Czaki) [reply napari_builtins/io/_read.py]: I'm not familiar with zarr arrays. I just follow the tests. And now waiting on new Zarr release.

File: napari/layers/labels/_tests/test_labels.py [hunk 2]
```diff
@@ -1110,6 +1112,11 @@ def test_large_label_values():
     assert len(np.unique(mapped.reshape((-1, 4)), axis=0)) == 4
 
 
+@pytest.mark.xfail(
+    parse_version(version('zarr')).is_prerelease,
+    reason='tensorstore is not compatible with zarr 3',
+    strict=True,
+)
```
Comment (jni) [inline napari/layers/labels/_tests/test_labels.py]: As I understand it, this is a problem with the zarr v3 format rather than zarr-python v3, in which case, perhaps we can change the test to write a zarr v2 file?

Comment (jni) [reply napari/layers/labels/_tests/test_labels.py]: https://zarr.readthedocs.io/en/v3/_autoapi/zarr/index.html#zarr.open

using `zarr_version=2` should work

**Comment by psobolewskiPhD** [comment]
a6 has https://github.com/zarr-developers/zarr-python/pull/2158, so we may be good to go?

**Comment by Czaki** [comment]
> a6 has [zarr-developers/zarr-python#2158](https://github.com/zarr-developers/zarr-python/pull/2158), so we may be good to go?

I will not have time before Thursday to sit on this.

**Comment by Czaki** [comment]
The `test_add_zarr_1d_array_is_ignored` and `test_add_many_zarr_1d_array_is_ignored` are failing because `__setitem__` for 1d array is not implemented. 
Did we really need to test if 1d arrays are ignored? 

# other failing are looking to need fixes on our side.

File: napari/layers/labels/_tests/test_labels.py (hunk 2 above)
Comment (Czaki) [reply napari/layers/labels/_tests/test_labels.py]: it looks working now

File: napari/layers/labels/_tests/test_labels.py [hunk 3]
```diff
@@ -1124,38 +1125,43 @@ def test_large_label_values():
...
 
 
-def test_fill_tensorstore():
+if parse_version(version('zarr')) > parse_version('3.0.0a0'):
+    driver = [(2, 'zarr'), (3, 'zarr3')]
+else:
+    driver = [(2, 'zarr')]
+
+
+@pytest.mark.parametrize(('zarr_version', 'zarr_driver'), driver)
```
Comment (jni) [inline napari/layers/labels/_tests/test_labels.py:1134]: 😍

File: napari_builtins/_tests/test_io.py [hunk 4]
```diff
@@ -68,23 +69,23 @@ def test_guess_zarr_path():
...
-        image_in = magic_imread([fout])
-        # Note: due to lazy loading, the next line needs to happen within
-        # the context manager. Alternatively, we could convert to NumPy here.
-        np.testing.assert_array_equal(image, image_in)
+    data_path = str(tmp_path / 'data.zarr')
+    z = zarr.open(store=data_path, mode='a', shape=image.shape)
+    z[:] = image
+    image_in = magic_imread([data_path])
+    # Note: due to lazy loading, the next line needs to happen within
+    # the context manager. Alternatively, we could convert to NumPy here.
```
Comment (jni) [inline napari_builtins/_tests/test_io.py]: There's no context manager anymore so we can remove this comment

```suggestion
```

File: napari_builtins/_tests/test_io.py [hunk 5]
```diff
@@ -104,24 +105,25 @@ def test_zarr_with_unrelated_file(tmp_path):
...
+
+    root = zarr.open_group(fout, mode='a')
+    for i in range(len(multiscale)):
+        shape = 20 // 2**i
+        z = root.create_dataset(str(i), shape=(shape,) * 2)
+        z[:] = multiscale[i]
+    multiscale_in = magic_imread([fout])
+    assert len(multiscale) == len(multiscale_in)
+    # Note: due to lazy loading, the next line needs to happen within
+    # the context manager. Alternatively, we could convert to NumPy here.
```
Comment (jni) [inline napari_builtins/_tests/test_io.py]: As above, no context manager = comment is outdated and can be removed.

```suggestion
```

File: napari_builtins/_tests/test_io.py [hunk 6]
```diff
@@ -312,28 +314,37 @@ def test_add_zarr(write_spec):
     assert out[0].shape == ZARR1.shape  # type: ignore
 
 
-def test_add_zarr_1d_array_is_ignored():
+@pytest.mark.skipif(
+    parse_version(version('zarr')) > parse_version('3a0'),
+    reason='zarr 3+ does not support setting 1d data',
```
Comment (jni) [inline napari_builtins/_tests/test_io.py]: Wait, sorry I missed your comment @Czaki, but I can't imagine this is actually true? @d-v-b got any insights as to why this fails now?

File: pyproject.toml [hunk 7]
```diff
@@ -362,7 +362,8 @@ filterwarnings = [
   # https://github.com/pydata/xarray/pull/8939
   "ignore:__array__ implementation doesn't accept a copy keyword, so passing copy=False failed.",
   "ignore:pkg_resources is deprecated",
-  "ignore:Deprecated call to `pkg_resources.declare_namespace"
+  "ignore:Deprecated call to `pkg_resources.declare_namespace",
+  "ignore:Use Group.create_array instead."
```
Comment (jni) [inline pyproject.toml:366]: Should this actually be ignored...? Maybe add a comment linking the relevant lines of code and any update guides?

File: resources/constraints/version_denylist.txt [hunk 8]
```diff
@@ -1,4 +1,3 @@
-matplotlib !=3.9.1 ; platform_system == 'Windows'
```
Comment (jni) [inline resources/constraints/version_denylist.txt:1]: Is this change intentional?

File: napari_builtins/io/_read.py [hunk 9]
```diff
@@ -133,6 +133,18 @@ def read_zarr_dataset(path: str):
         ]
         assert image, 'No arrays found in zarr group'
         shape = image[0].shape
+    elif (path / 'zarr.json').exists():
```
Comment (jni) [inline napari_builtins/io/_read.py:136]: We probably want to add a test for zarr v3? Or is this codecov being silly because only pre is testing v3 arrays?

File: resources/constraints/version_denylist.txt (hunk 8 above)
Comment (Czaki) [reply resources/constraints/version_denylist.txt:1]: yes. This release is yanked from PyPI

File: napari_builtins/_tests/test_io.py (hunk 6 above)
Comment (Czaki) [reply napari_builtins/_tests/test_io.py]: Zarr 2 silently drop 1d data (which is tested by this test). 

Zarr 3 raises `NotImplementedError`

File: napari_builtins/io/_read.py (hunk 9 above)
Comment (Czaki) [reply napari_builtins/io/_read.py:136]: yes. zarr 3 is only available in `--pre` tests

File: resources/constraints/version_denylist.txt (hunk 8 above)
Comment (jni) [reply resources/constraints/version_denylist.txt:1]: Ok. In general this should probably be a separate PR but let's ignore this time. 🙏

Comment (Czaki) [reply resources/constraints/version_denylist.txt:1]: I need to edit this file to trigger --pre tests

File: napari_builtins/_tests/test_io.py (hunk 6 above)
Comment (d-v-b) [reply napari_builtins/_tests/test_io.py]: i can't really tell what that test is doing -- is there a zarr bug, or is this "ignore 1D data" a napari thing?

Comment (jni) [reply napari_builtins/_tests/test_io.py]: No, this test checks that *we* don't *read* 1D data. But zarr (v2) is perfectly able to read and write 1D arrays.

```
In [1]: import zarr

In [2]: from tempfile import TemporaryDirectory

In [3]: td = TemporaryDirectory(suffix='.zarr')

In [4]: zarr.__version__
Out[4]: '2.18.3'

In [5]: td.name
Out[5]: '/var/folders/z7/mv1jznhx59q5vc3bp09gj69m0000gn/T/tmpcq58h89x.zarr'

In [6]: z = zarr.open(td.name, mode='w')

In [7]: arr = np.random.random(10)

In [8]: z['1d'] = arr

In [9]: arr2 = np.asarray(zarr.open(td.name, mode='r')['1d'])

In [10]: np.all(arr2 == arr)
Out[10]: np.True_
```

Comment (jni) [reply napari_builtins/_tests/test_io.py]: Looks like the issue is that creating an array with `__setitem__` on a **Group** is what doesn't work. I can make the 1D array with create_array, which is a bit clunky, but works:

```python
>>> from tempfile import TemporaryDirectory
>>> td = TemporaryDirectory(suffix='.zarr')
>>> import zarr
>>> zarr.__version__
'3.0.0b0'
>>> z = zarr.open(store=td.name, mode='w')
>>> import numpy as np
>>> arr = np.random.random(10)
>>> z['1d'] = arr
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/jni/micromamba/envs/z3/lib/python3.12/site-packages/zarr/core/group.py", line 1364, in __setitem__
    raise NotImplementedError
NotImplementedError
>>> type(z)
<class 'zarr.core.group.Group'>

>>> z.create_array(name='1d', shape=arr.shape, dtype=arr.dtype, data=arr)
<Array file:///var/folders/z7/mv1jznhx59q5vc3bp09gj69m0000gn/T/tmpcecbcl0a.zarr/1d shape=(10,) dtype=float64>
```

Comment (jni) [reply napari_builtins/_tests/test_io.py]: Oh LOL, just in time: zarr-developers/zarr-python#2393

Comment (d-v-b) [reply napari_builtins/_tests/test_io.py]: `create_array` allows you to specify the zarr array you are creating, whereas `z['id'] = my_numpy_array` does not, which makes the latter more magical, which is why we removed the "group is a dict" API in v3. is this a problem for napari?

Comment (d-v-b) [reply napari_builtins/_tests/test_io.py]: haha looks like I will need to bring this up in https://github.com/zarr-developers/zarr-python/pull/2393/files

Comment (Czaki) [reply napari_builtins/_tests/test_io.py]: @jni so why we expect `None` here 

```python
    assert npe2.read([image_path], stack=False) == [(None,)]
```

File: pyproject.toml (hunk 7 above)
Comment (Czaki) [reply pyproject.toml:366]: In this PR I try to make zarr test passing again, I do not fully understanad the test cases. 
```
napari_builtins/_tests/test_io.py:80 test_zarr_nested - DeprecationWarning: Use Group.create_array instead.…
napari_builtins/_tests/test_io.py F                                                                                                                                [ 4/83]
napari_builtins/_tests/test_io.py:91 test_zarr_with_unrelated_file - DeprecationWarning: Use Group.create_array instead.…
napari_builtins/_tests/test_io.py F                                                                                                                                [ 5/83]
napari_builtins/_tests/test_io.py:105 test_zarr_multiscale - DeprecationWarning: Use Group.create_array instead.…
```

File: resources/constraints/version_denylist.txt (hunk 8 above)
Comment (jni) [reply resources/constraints/version_denylist.txt:1]: It seems like we should have a better mechanism for that 😂 but ok!

Comment (Czaki) [reply resources/constraints/version_denylist.txt:1]: There is, but I prefer to validate it in another PR.

File: napari_builtins/_tests/test_io.py (hunk 6 above)
Comment (psobolewskiPhD) [reply napari_builtins/_tests/test_io.py]: napari won't open 1D anyways? If I try:
`viewer.add_image(np.ones(10))`
I get:
> ValueError: Image data must have at least 2 dimensions.

File: pyproject.toml (hunk 7 above)
Comment (psobolewskiPhD) [reply pyproject.toml:366]: Digging into zarr a bit, `create_dataset` was just for h5py compatibility.
See: https://github.com/zarr-developers/zarr-python/blob/a9d6d74c4a4d4efcdb64015919d0a5685a7e0239/src/zarr/core/group.py#L1630-L1646

And: https://github.com/zarr-developers/zarr-python/pull/571

As this is just in our tests, I think we can just replace `create_dataset` with `create_array` with no issues.

File: napari_builtins/_tests/test_io.py [hunk 10]
```diff
@@ -68,23 +69,21 @@ def test_guess_zarr_path():
...
 
 
 def test_zarr_nested(tmp_path):
     image = np.random.random((10, 20, 20))
     image_name = 'my_image'
     root_path = tmp_path / 'dataset.zarr'
-    grp = zarr.open(str(root_path), mode='a')
-    grp.create_dataset(image_name, data=image)
+    grp = zarr.open(store=str(root_path), mode='a')
+    grp.create_dataset(image_name, data=image, shape=image.shape)
```
Comment (psobolewskiPhD) [inline napari_builtins/_tests/test_io.py:86]:
```suggestion
    grp.create_array(image_name, data=image, shape=image.shape)
```
See: https://github.com/napari/napari/pull/7215/files#r1810721505

File: napari_builtins/_tests/test_io.py [hunk 11]
```diff
@@ -94,8 +93,8 @@ def test_zarr_with_unrelated_file(tmp_path):
...
```
Comment (psobolewskiPhD) [inline napari_builtins/_tests/test_io.py:97]:
```suggestion
    grp.create_array(image_name, data=image, shape=image.shape)
```
See: https://github.com/napari/napari/pull/7215/files#r1810721505

File: napari_builtins/_tests/test_io.py [hunk 12]
```diff
@@ -104,24 +103,23 @@ def test_zarr_with_unrelated_file(tmp_path):
...
-        # Note: due to lazy loading, the next line needs to happen within
-        # the context manager. Alternatively, we could convert to NumPy here.
-        for images, images_in in zip(multiscale, multiscale_in):
-            np.testing.assert_array_equal(images, images_in)
+    fout = str(tmp_path / 'multiscale.zarr')
+
+    root = zarr.open_group(fout, mode='a')
+    for i in range(len(multiscale)):
+        shape = 20 // 2**i
+        z = root.create_dataset(str(i), shape=(shape,) * 2)
```
Comment (psobolewskiPhD) [inline napari_builtins/_tests/test_io.py:117]:
```suggestion
        z = root.create_array(str(i), shape=(shape,) * 2)
```
See: https://github.com/napari/napari/pull/7215/files#r1810721505

File: napari_builtins/_tests/test_io.py (hunk 6 above)
Comment (psobolewskiPhD) [reply napari_builtins/_tests/test_io.py]: We skip 1D here:
https://github.com/Czaki/napari/blob/a451f643065ce5ce0242bf491cd76a2dbc90d1c9/napari_builtins/io/_read.py#L232-L234

So then None is returned

File: pyproject.toml (hunk 7 above)
Comment (psobolewskiPhD) [reply pyproject.toml:366]: I made that as a suggestion.

**Review by psobolewskiPhD — APPROVED** [review]
I tested locally and everything works as far as I can tell with a few simple zarr array tests -- there is no migration guide yet for v3.
I think the 1D thing doesn't really affect us, as we skip opening 1D arrays anyways.
To avoid the deprecation, should be no issue to just use `create_array` instead of `create_dataset`, because I don't think we care about h5py in our tests?

**Comment by Czaki** [comment]
> To avoid the deprecation, should be no issue to just use `create_array` instead of `create_dataset`, because I don't think we care about h5py in our tests?

Unfortunately, zarr 2.x does not have `create_array` method, so we need to keep `create_dataset` for zarr 2.x compatybility. See https://github.com/napari/napari/actions/runs/11660697193/job/32463676718?pr=7215#step:13:576

**Comment by psobolewskiPhD** [comment]
Gah, sorry! I've was using a zarr v3 env and reading those docs.
Annoying!

File: napari_builtins/_tests/test_io.py (hunk 6 above)
Comment (jni) [reply napari_builtins/_tests/test_io.py]: This test was added in #2743 to fix #1471. We should not ignore the test in v3, we should write out the 1D array and make sure again that it doesn't crash the viewer but is instead ignored.

I'll take a stab at a fix here using different code for different versions.

Comment (Czaki) [reply napari_builtins/_tests/test_io.py]: fixed tests

File: napari_builtins/io/_read.py [hunk 13]
```diff
@@ -133,6 +133,18 @@ def read_zarr_dataset(path: str):
...
+        # zarr v3
+        import zarr
+
+        data = zarr.open(store=path)
+        if isinstance(data, zarr.Array):
+            image = da.from_zarr(data)
+            shape = image.shape
+        else:
+            image = [data[k] for k in sorted(data)]
```
Comment (jni) [inline napari_builtins/io/_read.py:145]: @Czaki why do you think this works? Why is it different from the v2 version?

```python
            image = [
                read_zarr_dataset(subpath)[0]
                for subpath in sorted(path.iterdir())
                if not subpath.name.startswith('.') and subpath.is_dir()
            ]
```

Incidentally, while exploring this, I found that our current 1D skipping mechanism is completely broken, even in v2. If the *first* dataset is 1D, we skip the whole subdirectory. If a later dataset is 1D, we don't skip it at all.

Probably, we should merge this and fix that issue separately. Here's the diff to the test that fails, both here and in main:

```
@@ -356,3 +356,8 @@ def test_add_many_zarr_1d_array_is_ignored(tmp_path):
         else:
             assert isinstance(out[0], da.Array)
             assert out[0].ndim == int(name[0])
+    out1, out2, out3 = npe2.read([zarr_dir.as_posix()], stack=False)
+    assert out1 == (None,)
+    assert out2[0].ndim == 2
+    assert out3[0].ndim == 3
```

But anyway, to simplify flow between versions, I suggest making this the same as v2:

```suggestion
            image = [
                read_zarr_dataset(subpath)[0]
                for subpath in sorted(path.iterdir())
                if not subpath.name.startswith('.') and subpath.is_dir()
            ]
```

File: napari_builtins/_tests/test_io.py (hunk 12 above)
Comment (jni) [reply napari_builtins/_tests/test_io.py:117]: I guess we can use `.create` like @Czaki did in the test_1d functions?

Comment (Czaki) [reply napari_builtins/_tests/test_io.py:117]: I don't know. There is a lack of documentation. 
The `create` is also described as backward compatibility in v3.0 so still may emit some warning?

File: napari_builtins/io/_read.py (hunk 13 above)
Comment (Czaki) [reply napari_builtins/io/_read.py:145]: Here I use reading subarrays by zarr itself. 
At least zarr v3 read properly file with multiple groups.

File: napari_builtins/_tests/test_io.py [hunk 14]
```diff
@@ -312,28 +310,37 @@ def test_add_zarr(write_spec):
...
+    assert npe2.read([image_path], stack=False) == [(None,)]
 
 
-def test_add_many_zarr_1d_array_is_ignored():
+@pytest.mark.xfail(
+    parse_version(version('zarr')) >= parse_version('3.0.0a0')
+    and os.name == 'nt',
+    reason='zarr 3 return incorrect key in windows',
+    strict=True,
+)
```
Comment (Czaki) [inline napari_builtins/_tests/test_io.py:329]: I need to skip this test on windows because of 

https://github.com/zarr-developers/zarr-python/issues/2438

described here https://napari.zulipchat.com/#narrow/channel/212875-general/topic/handling.20zarr.203.2E0.20release/near/480752197

As it is strict `xfail` it will fail once zarr fix the bug.
